test/e2e/storage/testsuites/volume_modify.go

```diff
@@ -225,11 +222,209 @@ func (v *volumeModifyTestSuite) DefineTests(driver storageframework.TestDriver,
+
+		ginkgo.By("Modifying PVC to new valid VAC")
+		l.resource.Pvc = SetPVCVACName(ctx, l.resource.Pvc, l.vac.Name, f.ClientSet, setVACWaitPeriod)
+		gomega.Expect(l.resource.Pvc).NotTo(gomega.BeNil())
+
+		ginkgo.By("Checking PVC status")
+		err = e2epv.WaitForPersistentVolumeClaimModified(ctx, f.ClientSet, l.resource.Pvc, modifyVolumeWaitPeriod)
+		framework.ExpectNoError(err, "While waiting for PVC to have expected VAC")
+	})
+
+	ginkgo.It("should not create volume with VAC when csi driver does not support volume modification", func(ctx context.Context) {
+		if driver.GetDriverInfo().Capabilities[storageframework.CapModifyVolume] {
+			e2eskipper.Skipf("Driver %q supports VolumeAttributesClass - skipping negative test", driver.GetDriverInfo().Name)
+		}
+
+		l = local{}
+		l.config = driver.PrepareTest(ctx, f)
+		l.vac = MakeInvalidVAC(l.config)
+
+		ginkgo.By("Creating VolumeAttributesClass")
+		_, err := f.ClientSet.StorageV1beta1().VolumeAttributesClasses().Create(ctx, l.vac, metav1.CreateOptions{})
+		framework.ExpectNoError(err, "While creating VolumeAttributesClass")
+
+		ginkgo.By("Creating volume")
+		testVolumeSizeRange := v.GetTestSuiteInfo().SupportedSizeRange
+		l.resource = storageframework.CreateVolumeResourceWithVAC(ctx, driver, l.config, pattern, testVolumeSizeRange, &l.vac.Name)
+		ginkgo.DeferCleanup(cleanup)
+
+		ginkgo.By("Creating a pod with dynamically provisioned volume")
+		podConfig := e2epod.Config{
+			NS:            f.Namespace.Name,
+			PVCs:          []*v1.PersistentVolumeClaim{l.resource.Pvc},
+			SeLinuxLabel:  e2epod.GetLinuxLabel(),
+			NodeSelection: l.config.ClientNodeSelection,
+			ImageID:       e2epod.GetDefaultTestImageID(),
+		}
+		pod, err := e2epod.CreateSecPodWithNodeSelection(ctx, f.ClientSet, &podConfig, f.Timeouts.PodStart)
+		ginkgo.DeferCleanup(e2epod.DeletePodWithWait, f.ClientSet, pod)
+		framework.ExpectNoError(err, "While creating test pod with VAC")
+
+		ginkgo.By("Checking events and PVC status")
+		eventSelector := fields.Set{
+			"involvedObject.kind":      "PersistentVolumeClaim",
+			"involvedObject.name":      l.resource.Pvc.Name,
+			"involvedObject.namespace": l.resource.Pvc.Namespace,
+			"reason":                   events.ProvisioningFailed,
+		}.AsSelector().String()
+		msg := "controller MODIFY_VOLUME capability is not reported"
+
+		err = e2eevents.WaitTimeoutForEvent(ctx, f.ClientSet, f.Namespace.Name, eventSelector, msg, f.Timeouts.ClaimProvision)
+		// Events are unreliable, don't depend on the event. It's used only to speed up the test.
+		if err != nil {
+			framework.Logf("Warning: did not get event about %s", events.ProvisioningFailed)
+		}
+
+		pvc, err := f.ClientSet.CoreV1().PersistentVolumeClaims(l.resource.Pvc.Name).Get(ctx, l.resource.Pvc.Name, metav1.GetOptions{})
+		framework.ExpectNoError(err, "could not re-read the pvc after event (or timeout)")
+		gomega.Expect(pvc.Status.Phase).To(gomega.Equal(v1.ClaimPending), "PVC phase isn't pending")
+	})
+
```

Comment:
Missing roll-back recovery test case "should recover from invalid target VAC by updating PVC to previous VAC". See [this comment](https://github.com/kubernetes/kubernetes/pull/126423#issuecomment-2258434420). Will raise issue and followup PR with that test later.